corehq/form_processor/backends/sql/dbaccessors.py

```diff
@@ -407,6 +437,40 @@ def delete_all_cases(domain=None):
             cursor.execute('SELECT delete_all_cases(%s)', [domain])
 
     @staticmethod
+    def get_all_cases_modified_since(server_modified_on_since=None, chunk_size=500):
+        start_from = server_modified_on_since or datetime.min
+        # todo: this will greedily query the same cases multiple times in a sharded
+        # setup. We should make it smarter
+        batch = CaseAccessorSQL.get_cases_modified_since(start_from, limit=chunk_size)
+        while len(batch) == chunk_size:
+            for case in batch:
+                yield case
+                next_start_from = case.server_modified_on
+
+            # make sure we are making progress
+            assert next_start_from > start_from
+            start_from = next_start_from
+            batch = CaseAccessorSQL.get_cases_modified_since(start_from, limit=chunk_size)
+
+        # last batch
+        for case in batch:
+            yield case
+
+    @staticmethod
+    def get_cases_modified_since(server_modified_on_since=None, limit=500):
+        """
+        Iterate through all cases in the entire database, optionally modified since
+        a specific date
+        """
+        if server_modified_on_since is None:
+            server_modified_on_since = datetime.min
+        results = list(CommCareCaseSQL.objects.raw('SELECT * FROM get_all_cases_modified_since(%s, %s)',
+                                                [server_modified_on_since, limit]))
```

Comment:
Good point about `opened_on` being a phone date. Doesn't seem necessary to
add it just for this one use case.

On 11 February 2016 at 09:04, Cory Zue <notifications@github.com> wrote:

> In corehq/form_processor/backends/sql/dbaccessors.py
> <https://github.com/dimagi/commcare-hq/pull/10259#discussion_r52568808>:
>
> > +            batch = CaseAccessorSQL.get_cases_modified_since(start_from, limit=chunk_size)
> > +
> > +        # last batch
> > +        for case in batch:
> > +            yield case
> > +
> > +    @staticmethod
> > +    def get_cases_modified_since(server_modified_on_since=None, limit=500):
> > +        """
> > +        Iterate through all cases in the entire database, optionally modified since
> > +        a specific date
> > +        """
> > +        if server_modified_on_since is None:
> > +            server_modified_on_since = datetime.min
> > +        results = list(CommCareCaseSQL.objects.raw('SELECT * FROM get_all_cases_modified_since(%s, %s)',
> > +                                                [server_modified_on_since, limit]))
>
> I think the most important criteria is that once we start iterating
> through dates, nothing can ever get inserted into the data that we've
> already seen. So from that perspective server_modified_on is ok, though
> could result in some cases being returned multiple times if they are
> modified while we're iterating. I don't think we can use opened_on
> because that's a phone date which could be in the past. If we had
> server_opened_on that would probably be best (should we consider adding
> that to the model?)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/10259/files#r52568808>.
>



-- 
Simon Kelly
Senior Engineer | Dimagi South Africa